src/Core/Framework/App/Url/AppUrlVerifier.php

```diff
@@ -0,0 +1,273 @@
+    {
+        $lock = $this->lockFactory->createLock('app-url-verification', 3);
+
+        try {
+            if ($lock->acquire()) {
+                return $lock;
+            }
+        } catch (LockConflictedException|LockAcquiringException) {
+        }
+
+        return null;
+    }
+
+    private function performVerification(string $appUrl, int $tries = 1): VerificationState
+    {
+        [$status, $info] = $this->executeVerify($appUrl);
+
+        $state = new VerificationState($status, $tries, $this->clock->now(), $info);
+
+        $this->persist($state);
+
+        return $state;
+    }
+
+    private function persist(VerificationState $state): void
+    {
+        $item = $this->cache->getItem(self::VERIFICATION_RESULT_CACHE_KEY);
+        $item->set($state);
+
+        if ($state->is(VerificationStatus::HARD_FAIL)) {
+            $item->expiresAt(null); // never expire
+        } else {
+            $item->expiresAfter(60 * 60 * 24); // 24h
+        }
+
+        $this->cache->save($item);
+    }
+
+    /**
+     * @return array{0: VerificationStatus, 1: string|null}
+     */
+    private function executeVerify(string $appUrl): array
+    {
+        try {
+            $url = $this->buildVerificationUrl($appUrl);
+        } catch (AppException $e) {
+            return [VerificationStatus::HARD_FAIL, $e->getMessage()];
+        }
+
+        $runId = bin2hex(random_bytes(8));
+        $cacheKey = self::VERIFICATION_CACHE_KEY_PREFIX . $runId;
+        $token = $this->createAndPersistToken($cacheKey);
+
+        try {
+            $resp = $this->httpClient->request(
+                'GET',
+                $url,
+                $this->buildRequestOptions($runId, $token)
+            );
+
+            if ($resp->getStatusCode() === Response::HTTP_NO_CONTENT) {
+                return [VerificationStatus::PASS, null];
+            }
+
+            if ($resp->getStatusCode() >= Response::HTTP_INTERNAL_SERVER_ERROR) {
+                return [VerificationStatus::SOFT_FAIL, $resp->getContent(false)];
+            }
+        } catch (TransportExceptionInterface $e) {
+            return [VerificationStatus::SOFT_FAIL, 'Failed to connect to APP_URL'];
+        } catch (\Throwable $e) {
+            return [VerificationStatus::HARD_FAIL, $e->getMessage()];
+        } finally {
+            $this->cache->deleteItem($cacheKey);
+        }
+
+        return [VerificationStatus::HARD_FAIL, 'Verification failed'];
+    }
+
+    /**
+     * @return array<string, mixed>
+     */
+    private function buildRequestOptions(string $runId, string $token): array
+    {
+        return [
+            'max_redirects' => 3,
+            'timeout' => 2.0,
+            'connect_timeout' => 0.5,
+            'query' => ['rid' => $runId, 'token' => $token],
+            'headers' => [
+                'Cache-Control' => 'no-store, no-cache, max-age=0',
+                'Pragma' => 'no-cache',
+                'User-Agent' => 'Shopware-AppUrlVerifier',
```

Comment:
can we append here the used shopware version, that might get helpful in future

Reply:
Sure!